robots/pr-labeler/main.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Commenter provides a way to --query for issues and append a --comment to matches.
+//
+// The --token determines who interacts with github.
+// By default commenter runs in dry mode, add --confirm to make it leave comments.
+// The --updated, --include-closed, --ceiling options provide minor safeguards
+// around leaving excessive comments.
+package main
+
+import (
+	"flag"
+	"log"
+	"net/url"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/flagutil"
+	"k8s.io/test-infra/prow/github"
+)
+
+const (
+	needsOkToTest = "needs-ok-to-test"
+	okToTest      = "ok-to-test"
+)
+
+type client interface {
+	AddLabel(org, repo string, number int, label string) error
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	GetPullRequests(org, repo string) ([]github.PullRequest, error)
+	ListOrgMembers(org, role string) ([]github.TeamMember, error)
+}
+
+type options struct {
+	confirm  bool
+	endpoint flagutil.Strings
+	org      string
+	repo     string
+	token    string
+}
+
+func flagOptions() options {
+	o := options{
+		endpoint: flagutil.NewStrings("https://api.github.com"),
+		org:      "kubernetes",
+		repo:     "test-infra",
+	}
+	flag.BoolVar(&o.confirm, "confirm", false, "Mutate github if set")
+	flag.StringVar(&o.token, "token", "", "Path to github token")
+	flag.Parse()
+	return o
+}
+
+func main() {
+	log.SetFlags(log.LstdFlags | log.Lshortfile)
+	o := flagOptions()
+
+	if o.token == "" {
+		log.Fatal("empty --token")
+	}
+
+	secretAgent := &config.SecretAgent{}
+	if err := secretAgent.Start([]string{o.token}); err != nil {
+		log.Fatalf("Error starting secrets agent: %v", err)
+	}
+
+	var err error
+	for _, ep := range o.endpoint.Strings() {
+		_, err = url.ParseRequestURI(ep)
+		if err != nil {
+			log.Fatalf("Invalid --endpoint URL %q: %v.", ep, err)
+		}
+	}
+
+	var c client
+	if o.confirm {
+		c = github.NewClient(secretAgent.GetTokenGenerator(o.token), o.endpoint.Strings()...)
+	} else {
+		c = github.NewDryRunClient(secretAgent.GetTokenGenerator(o.token), o.endpoint.Strings()...)
+	}
+
+	// get all open PRs
+	prs, err := c.GetPullRequests(o.org, o.repo)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// get the list of members once and use a set for lookups
+	m, err := c.ListOrgMembers(o.org, "member")
```

Comment:
Is this identical with the logic now? I think it's repo collaborators?

Reply:
Actually we check:
1. ghc.IsCollaborator(org, repo, user)
1. ghc.IsMember(org, user)
1. ghc.IsMember(trigger.TrustedOrg, user)

I guess the collaborators include members?
It's true that I am not checking for another trusted org... at worse we'll have some trusted PRs with `ok-to-test` labels, which doesn't prevent tests from being run.

Reply:
Ok, but actually I am skipping PRs from collaborators, so we might end up with PRs that are double-trusted because:
1. they have `ok-to-test`
1. they are authored by a collaborator

I will still fix that for the sake of correctness :)